### PR TITLE
Adds PL Visualizations Context Manager [WIP]

### PR DESCRIPTION
**Pull request by Lucaman99**
**Context:**

Adds a new PennyLane visualization context manager, which allows for the visualization of variational quantum circuits.

giphy

**Description of the Change:**

If a user wishes to visualize the optimization of their variational quantum circuit, they simply have to wrap the optimization loop in the `Visualize` context manager:

```python
from pennylane.visualize import Visualize

steps = 100

def cost(params):
     # Cost function here

with Visualize(steps, cost_fn=cost) as viz:
      for i in range(steps):
          # Optimization here
```

The `Visualize` context manager records the data (value of the cost function and value of the variational parameters) associated with each optimization step. To record data for each step, the user must call `viz.update(params=params)` inside the optimization procedure:

```python
from pennylane.visualize import Visualize

steps = 100
optimizer = qml.GradientDescentOptimizer()

def cost(params):
     # Cost function here

with Visualize(steps, cost_fn=cost) as viz:
      for i in range(steps):
           params = optimizer.step(cost, params)
           viz.update(params=params)
```

This recorded data can be accessed by the user after the optimization is complete, or it can be used to "visualize" each optimization step. The user can call `viz.text()` to display a text-based output with the data associated with each step, as well as `viz.graph()` either inside or outside the optimization loop to dynamically print a graph as the cost function optimizes, or to print the final graph of the cost function after the optimization has concluded.

```python
from pennylane.visualize import Visualize

steps = 100
optimizer = qml.GradientDescentOptimizer()

def cost(params):
     # Cost function here

with Visualize(steps, cost_fn=cost) as viz:
      for i in range(steps):
           params = optimizer.step(cost, params)
           viz.update(params=params)
           viz.text()
           viz.graph()
```

This is still a WIP PR, we still have to:

- Write tests
- Finish writing the docstrings
- Ensure that all error handling is robust

**Benefits:**

Will allow users to visualize variational quantum circuits in an intuitive and visually-pleasing way.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: Untitled.ipynb [hunk 1]
```diff
@@ -0,0 +1,190 @@
+{
```
Comment (ixfoduap) [inline Untitled.ipynb:1]: I think this file should be removed, right?

File: pennylane/visualize/visualize.py [hunk 2]
```diff
@@ -0,0 +1,351 @@
+# Copyright 2020 Xanadu Quantum Technologies Inc.
+
```
Comment (ixfoduap) [inline pennylane/visualize/visualize.py:2]: I know this is a WIP, bit it's best to handle these codefactor issues from the beginning!

File: pennylane/visualize/visualize.py [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+    """
+    A context manager for storing data relating to visualizations and post-processing of
+    variational quantum circuit optimization.
+
+    When a PennyLane optimization loop is wrapped in the Visualize context manager, data from each optimization step
+    is recorded and outputted as text or a graph.
+
+    Args:
+        steps (int): The number of steps for which the circuit is being optimized.
```
Comment (ixfoduap) [inline pennylane/visualize/visualize.py:43]: I wonder why this is needed? The UI is still to run a separate loop for optimization. I tried locally to run the code 

```python
with Visualize(10, cost_fn=cost) as viz:
    for i in range(50):
        params = opt.step(cost, params)
```
and it works fine, except the print statements are wrong, e.g., 50 / 10

**Review by ixfoduap — COMMENTED** [review]
I haven't reviewed the code yet, but instead have been playing with the visualization on a jupyter notebook. Some comments:

1. When I only print the text or the cost function, things work well. But when both are printed at the same time, weird things happen: the text gets erased and there is "blinking" with each step that is very annoying
2. For a small example, the time taken to build the plot is slower than to compute one step of the optimization, so optimization slows down when we visualize it! This isn't good and we should do something about it
3. I think it would be best to define the plot with the x-axis already covering all the intended steps of the optimization. This way we can see how optimization _progresses_ . Right now it's as if a totally new plot is created

Still a lot of work left, but this is going to be great once it's finished!

File: pennylane/visualize/visualize.py (hunk 2 above)
Comment (Lucaman99) [reply pennylane/visualize/visualize.py:2]: Absolutely, I'll fix this right away

File: Untitled.ipynb (hunk 1 above)
Comment (Lucaman99) [reply Untitled.ipynb:1]: Yes, that was a mistake, will remove it!

**Comment by glassnotes** [comment]
@Lucaman99 this is a very helpful tool! I look forward to it being incorporated in PennyLane :rocket: 

For context, I'm currently working on comparing a handful of different optimizers for a VQE problem. After trying it out in that setting, I have a couple suggestions:
 - I ran into some trouble when trying to output both text and plot. Is it possible to have the plot stay consistently at the top, and the text continuously update underneath?  (I've included a gif below).
 - Can timing information and number of device executions be added to allow one to track the time of each iteration / total optimization? This isn't necessarily "visualization" but since this is wrapping the optimization it would be a convenient place to put it (plus then you can make graphs of cost vs. time, cost vs. executions, etc.)
 - It would be cool to be able to compare multiple optimizers on the same plot (e.g. pass a list of params for each optimizer, and have them plotted as separate curves) 
 - Finally this is a minor suggestion but would love to be able to pass `colour=some_colour` to the `viz.graph()` function.  (This would be an especially nice addition if multiple optimizer support could be added, so that one could explicitly specify colours for them)

Gif of the text+graph behaviour:
viz

File: pennylane/visualize/visualize.py [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+        self._step_log += 1
+
+        if self._step_log % self.step_size == 0:
+
+            self._x_log.append(self._step_log)
+
+            if params is not None:
+                self._param_log.append(params)
+                if self.cost_fn is not None:
+                    val = self.cost_fn(params)
```
Comment (josh146) [inline pennylane/visualize/visualize.py:163]: @Lucaman99 note that there is now an `optimizer.step_and_cost()` function, which returns the parameter update _and_ the cost function value in a single quantum evaluation. This will avoid doubling the number of quantum evaluations occurring.

**Comment by josh146** [comment]
@Lucaman99 @ixfoduap I might close this PR for now -- feel free to re-open if work resumes!
